board/safety/safety_tesla.h

```diff
@@ -0,0 +1,319 @@
+// board enforces
+//   in-state
+//      accel set/resume
+//   out-state
+//      cancel button
+//      regen paddle
+//      accel rising edge
+//      brake rising edge
+//      brake > 0mph
+
+// 2m/s are added to be less restrictive
+const struct lookup_t TESLA_LOOKUP_ANGLE_RATE_UP = {
+    {2., 7., 17.},
+    {5., .8, .25}};
+
+const struct lookup_t TESLA_LOOKUP_ANGLE_RATE_DOWN = {
+    {2., 7., 17.},
+    {5., 3.5, .8}};
+
+const struct lookup_t TESLA_LOOKUP_MAX_ANGLE = {
+    {2., 29., 38.},
+    {410., 92., 36.}};
+
+const int TESLA_RT_INTERVAL = 250000; // 250ms between real time checks
+
+struct sample_t tesla_angle_meas; // last 3 steer angles
+
+// state of angle limits
+int tesla_desired_angle_last = 0; // last desired steer angle
+int16_t tesla_rt_angle_last = 0.; // last real time angle
+uint32_t tesla_ts_angle_last = 0;
+
+int tesla_controls_allowed_last = 0;
+int steer_allowed = 1;
+
+int tesla_brake_prev = 0;
+int tesla_gas_prev = 0;
+int tesla_speed = 0;
+int current_car_time = -1;
+int time_at_last_stalk_pull = -1;
+int eac_status = 0;
+
+int tesla_ignition_started = 0;
+
+static void tesla_rx_hook(CAN_FIFOMailBox_TypeDef *to_push)
+{
+  set_gmlan_digital_output(GMLAN_HIGH);
+  reset_gmlan_switch_timeout(); //we're still in tesla safety mode, reset the timeout counter and make sure our output is enabled
+
+  //int bus_number = (to_push->RDTR >> 4) & 0xFF;
+  uint32_t addr;
+  if (to_push->RIR & 4)
+  {
+    // Extended
+    // Not looked at, but have to be separated
+    // to avoid address collision
+    addr = to_push->RIR >> 3;
+  }
+  else
+  {
+    // Normal
+    addr = to_push->RIR >> 21;
+  }
+
+  // Record the current car time in current_car_time (for use with double-pulling cruise stalk)
+  if (addr == 0x318)
+  {
+    int hour = (to_push->RDLR & 0x1F000000) >> 24;
+    int minute = (to_push->RDHR & 0x3F00) >> 8;
+    int second = (to_push->RDLR & 0x3F0000) >> 16;
+    current_car_time = (hour * 3600) + (minute * 60) + second;
```

Comment:
using car time from CAN seems unnecessary complicated and potentially unreliable. I recommend re-using:
```
    uint32_t ts = TIM2->CNT;
    uint32_t ts_elapsed = get_ts_elapsed(ts, tesla_ts_angle_last);
```

Reply:
I got rid of this section of code.  It was going to be used if we only used double-pulls to activate openpilot.  We now are activating on a single pull.  Double-pull may activate other features, but safety-wise, we only need to monitor a single pull because nothing changes from a safety perspective when you pull twice.

Reply:
cool